### ISSUE TITLE
[analyzer] Failed iostream state is not constrained: `if (is)` is taken after a failed `getline`/`operator>>`/`setstate`

### ISSUE THREAD

**Issue by TedLyngmo** [llvm/llvm-project#226206]
## Summary

The static analyzer treats a contradictory `std::basic_ios` state as reachable: after an operation on a stream is known to have failed, a subsequent `if (is)` still splits the state and the "stream is good" branch is explored. As a consequence, any code that reports failure through stream state — the idiom the standard itself uses for `operator>>` — produces false positives from the uninitialized-value checkers.

## Reproducer 1 — the contradiction, isolated

Every null dereference below sits on a provably unreachable line. Each is guarded by `if (is)` on a stream whose immediately preceding operation is known to have failed, with no intervening `clear()`.

```cpp
#include <istream>
#include <string>

void three_arg(std::istream& is)
{
    std::string s;
    if (std::getline(is, s, '\0')) { return; }    // getline failed past this point
    if (is) { *static_cast<int*>(nullptr) = 1; }  // warning: Dereference of null pointer
}

void two_arg(std::istream& is)
{
    std::string s;
    if (std::getline(is, s)) { return; }
    if (is) { *static_cast<int*>(nullptr) = 1; }  // warning: Dereference of null pointer
}

void setstate(std::istream& is)
{
    is.setstate(std::ios::failbit);
    if (is) { *static_cast<int*>(nullptr) = 1; }  // warning: Dereference of null pointer
}
```

```
clang++ --analyze -std=c++17 -Xanalyzer -analyzer-checker=core contradiction.cpp
```

All three are reported. `is.get()` and `is >> value` behave the same way as `getline`.

## Reproducer 2 — the practical consequence

An `operator>>` that leaves its out-parameter alone and sets `failbit` on bad input, with a caller that aborts when the extraction fails. The field is written on every path that reaches the end of the constructor, yet two checkers fire.

```cpp
#include <cstdlib>
#include <fstream>
#include <istream>
#include <string>

enum class Product { A, B };

inline std::istream& operator>>(std::istream& is, Product& product)
{
    if (std::string name; std::getline(is, name, '\0')) {
        if (name == "A")      { product = Product::A; }
        else if (name == "B") { product = Product::B; }
        else                  { is.setstate(std::ios::failbit); }
    }
    return is;
}

struct Info {
    Info()
    {
        std::ifstream rev("/some/file");
        if (not rev) { std::abort(); }
        env = 1;
        if (not(rev >> product)) { std::abort(); }
    }  // warning: 1 uninitialized field at the end of the constructor call
    int env;
    Product product;
};

int sink(Product p);
int use() { return sink(Info{}.product); }  // warning: uninitialized value
```

```
clang++ --analyze -std=c++17 -Xanalyzer -analyzer-output=text \
    -Xanalyzer -analyzer-checker=core,optin.cplusplus.UninitializedObject \
    extraction_operator.cpp
```

```
warning: 1 uninitialized field at the end of the constructor call [optin.cplusplus.UninitializedObject]
warning: 1st function call argument is an uninitialized value [core.CallAndMessage]
  note: Calling default constructor for 'Info'
  note: Taking false branch                    <- ifstream opened successfully
  note: Calling 'operator>>'
  note: Taking false branch                    <- getline FAILED, product not written
  note: Returning from 'operator>>'
  note: Taking false branch                    <- ...yet the stream tests good, so no abort()
  note: 1st function call argument is an uninitialized value
```

The path is self-contradictory: `getline` failed, so `rev` must be in a failed state, so `not(rev >> product)` must be true and `std::abort()` must run. The analyzer takes the opposite branch and reaches the end of the constructor with `product` unwritten.

## Why this matters

Reporting extraction failure via stream state, leaving the out-parameter untouched, is what [istream.extractors] prescribes and what every user-defined `operator>>` does. Under these checkers that idiom is unusable: the only ways to silence the report are to add a redundant
initialization that no correct path can observe, to restructure so failure is carried in a return value instead of the stream, or to suppress. None of those are improvements to the code.

## Versions

Reproduces with all three functions of reproducer 1 reported, unchanged, on:

| clang | result |
|-------|--------|
| 14.0.6 | 3/3 |
| 15.0.5 | 3/3 |
| 16.0.3 | 3/3 |
| 17.0.6 | 3/3 |
| 19.1.6 | 3/3 |
| [trunk (via Compiler Explorer)](https://godbolt.org/z/YM3z6dnPe) | 3/3 |

**Comment by NagyDonat** [comment]
As of now, the static analyzer does not have any code for modeling the behavior of `iostream` objects.

The data model of the analyzer is fairly low-level and does not offer good tools for modeling C++ objects, so it is very cumbersome to develop code that "teaches" the analyzer the behavior and invariants of a C++ class (or class hierarchy) like `iostream`.

We do have some code that models the behavior of smart pointers and C++ iterators, but those are infamously complex and unstable. (Personally I think that their approach is a dead-end, they should be rewritten from scratch after developing a proper framework for modeling objects in the analyzer engine. Needless to say, developing such framework is a _very_ big task.)

If there number of people working on the analyzer increased significantly, they would eventually develop support for C++ STL classes like `iostream`, which would resolve this issue. However, in the foreseeable future I don't see a chance that there would be progress on this ticket, so I'm closing it as "not planned".
